### ISSUE TITLE
questions about initializing the lstm hidden states

### ISSUE THREAD

**Issue by brisker** [ruotianluo/ImageCaptioning.pytorch#25]
here :https://github.com/ruotianluo/neuraltalk2.pytorch/blob/master/models/OldModel.py#L49
you seems to directly init the hidden states with the fc_feats with a linear layer. So I want to ask that if I want to implement an **attention model** where the lstm takes **fc_feats as input** **at step 0**,  and takes start token as input at step 1, like the figure below, then how to init the hidden states of lstm?
![image](https://user-images.githubusercontent.com/13804492/32820800-3cd22af8-ca0b-11e7-84b7-ca558ec67d55.png)

**Comment by ruotianluo** [comment]
Isn't show attend tell in the same file the one you want?

**Comment by brisker** [comment]
@ruotianluo 
emm, I know this is show _attend _tell model. But it seems that this implementation just use fc_feats to init the hidden states, not to use it as input as step 0. At step 0 , you seems to directly output a word . But at step 0 I just wanna output a start token and do not include attention temporarily.

**Comment by ruotianluo** [comment]
Intialize with zero.

**Comment by brisker** [comment]
@ruotianluo 
here: https://github.com/ruotianluo/neuraltalk2.pytorch/blob/master/train.py#L113
why `labels[:,1:], masks[:,1:]`? not `labels[:,:], masks[:,:]` ?

**Comment by ruotianluo** [comment]
The zero-th are bos.

**Comment by brisker** [comment]
@ruotianluo 
what is bos.....

**Comment by ruotianluo** [comment]
start token.

**Comment by brisker** [comment]
@ruotianluo 
If so, what is the function of `masks`?

**Comment by brisker** [comment]
@ruotianluo 
if output is [0, 12, 20,0,0] (1 is start token and 0 is end token ), then
masks should be [1,1,1,0,0]?

**Comment by ruotianluo** [comment]
mask would be 1,1,1,1,0

**Comment by brisker** [comment]
@ruotianluo  
still:
target is [1, 12, 20,0,0] (1 is start token and 0 is end token )
and if output of rnn is [1 ,12,21,0,0], do I need to write like:
`crit(RNNDecoder_output[:,1:], labels[:, 1:], masks[:, 1:])`
or
`crit(RNNDecoder_output[:,:], labels[:, 1:], masks[:, 1:])`

**Comment by ruotianluo** [comment]
Outputs are softmax probabilities.

**Comment by brisker** [comment]
@ruotianluo 
how to add weights to different word?
suppose in target [1, 12,20,0,0], I want to assign 12 with weight 5, and assign 20 with weight 1?

**Comment by ruotianluo** [comment]
You have to modify the crit. It's nontrivial.

**Comment by brisker** [comment]
@ruotianluo 
is changing mask from [1 1 1 1 0] to []1 5 1 1 0] not correct?

**Comment by ruotianluo** [comment]
I see what you mean. I misunderstood.  Yeah, it should work.

**Comment by brisker** [comment]
@ruotianluo
thanks. further question... I am new to caption. So have you seen any papers that do weighted loss or different losses on image caption task? If yes, why may we need these losses? Can you provide any simple overview?

**Comment by brisker** [comment]
@ruotianluo 
here:https://github.com/ruotianluo/ImageCaptioning.pytorch/blob/master/models/ShowTellModel.py#L79
it seems that in your implementation you use the output of lstm as word prediction. Why not directly using 
hidden states as word prediction?
